### DISCUSSION TITLE
Hardware acceleration for 4K

### DISCUSSION THREAD

**Discussion by pixel24** [jellyfin/jellyfin#9774] [General]
Hi@all,

I have to admit that I'm not that knowledgeable about video codec etc.. I ask for a little indulgence :-)

I am using Jellyfin (version: 10.8.10) on an Ubuntu server (23.04) running as a VM on Proxmox. Additionally I have installed a nVIDIA card in the server and passed it through to the VM:

```
root@media01:~# lspci -nn | grep -Ei "3d|display|vga"
00:01.0 VGA compatible controller [0300]: Device [1234:1111] (rev 02)
01:00.0 VGA compatible controller [0300]: NVIDIA Corporation GP107GL [Quadro P620] [10de:1cb6] (rev a1)
```
media01 ist die Ubuntu-VM. The correct version of ffmpeg is also installed:

```
root@media01:~# /usr/lib/jellyfin-ffmpeg/ffmpeg 
ffmpeg version 5.1.3-Jellyfin Copyright (c) 2000-2022 the FFmpeg developers
  built with gcc 11 (Ubuntu 11.3.0-1ubuntu1~22.04)
  configuration: --prefix=/usr/lib/jellyfin-ffmpeg --target-os=linux --extra-libs=-lfftw3f --extra-version=Jellyfin --disable-doc --disable-ffplay --disable-ptx-compression --disable-static --disable-libxcb --disable-sdl2 --disable-xlib --enable-lto --enable-gpl --enable-version3 --enable-shared --enable-gmp --enable-gnutls --enable-chromaprint --enable-libdrm --enable-libass --enable-libfreetype --enable-libfribidi --enable-libfontconfig --enable-libbluray --enable-libmp3lame --enable-libopus --enable-libtheora --enable-libvorbis --enable-libopenmpt --enable-libdav1d --enable-libwebp --enable-libvpx --enable-libx264 --enable-libx265 --enable-libzvbi --enable-libzimg --enable-libfdk-aac --arch=amd64 --enable-libsvtav1 --enable-libshaderc --enable-libplacebo --enable-vulkan --enable-opencl --enable-vaapi --enable-amf --enable-libmfx --enable-ffnvcodec --enable-cuda --enable-cuda-llvm --enable-cuvid --enable-nvdec --enable-nvenc
  libavutil      57. 28.100 / 57. 28.100
  libavcodec     59. 37.100 / 59. 37.100
  libavformat    59. 27.100 / 59. 27.100
  libavdevice    59.  7.100 / 59.  7.100
  libavfilter     8. 44.100 /  8. 44.100
  libswscale      6.  7.100 /  6.  7.100
  libswresample   4.  7.100 /  4.  7.100
  libpostproc    56.  6.100 / 56.  6.100
Hyper fast Audio and Video encoder
usage: ffmpeg [options] [[infile options] -i infile]... {[outfile options] outfile}
```
I still have NGINX with LE certificates on the Jellyfin server. If these configurations are relevant I can also post them.

I set the settings for hardware acceleration in Jellyfin as follows:

![grafik](https://github.com/jellyfin/jellyfin/assets/16030132/3467367b-1992-490b-bdc1-6fe6a2e7934c)

For this test I use the following film:

![grafik](https://github.com/jellyfin/jellyfin/assets/16030132/6f14b300-36ce-4732-aaca-e43b46a47cf3)

When I watch it on my computer (HP Workstation Z440) via the Jellyfin web interface, I only see minimal judder. Really little. The CPU load of the VM increases to about 12%. that looks good (I think).

However, the primary client is a Zotac (mini-computer) on the TV with libreelec + Kodi PlugIn. Ok, this client is not the fastest and should be replaced soon:

```
tvbo001:~ # cat /proc/cpuinfo 
processor       : 0
vendor_id       : GenuineIntel
cpu family      : 6
model           : 58
model name      : Intel(R) Core(TM) i3-3220T CPU @ 2.80GHz
stepping        : 9
microcode       : 0x21
cpu MHz         : 1600.000
cache size      : 3072 KB
physical id     : 0
siblings        : 4
core id         : 0
cpu cores       : 2
apicid          : 0
initial apicid  : 0
fpu             : yes
fpu_exception   : yes
cpuid level     : 13
wp              : yes
flags           : fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl est tm2 ssse3 cx16 xtpr pdcm pcid sse4_1 sse4_2 popcnt tsc_deadline_timer xsave avx f16c lahf_lm cpuid_fault epb pti ssbd ibrs ibpb stibp fsgsbase smep erms xsaveopt dtherm arat pln pts md_clear flush_l1d
bugs            : cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs itlb_multihit mmio_unknown
bogomips        : 5589.10
clflush size    : 64
cache_alignment : 64
address sizes   : 36 bits physical, 48 bits virtual
power management:
..
```
However, I thought if the server takes care of the encoding, the client doesn't have to be so powerful. Or have I misunderstood?
If I start this movie from Kodi, the CPU load of the server VM remains almost at 0%. Whereas the CPU load on the client (Zotac computer) is completely at its limit.

It seems to me that hardware acceleration is not being used when playing through Kodi. Or do I have an error in the configuration?

with best

**Comment by Shadowghost** [comment]
Main culprit is most likely Kodi because it requests the bare file and uses the local hardware to decode it. Which is problematic in your case because a 3rd gen i3 is not capable to decode or transcode HEVC or HDR. If you use Jellycon you can force a transcode and set the appropriate codec options within the settings.

Additionally you should enable tonemapping (if your GPU supports it) otherwise the colors on transcodes of HDR content will look washed out if.

**Comment by pixel24** [comment]
The fact that the hardware acceleration in the server is not used via the "normal" Kodi Jellyfin plugin is not nice :-(

Does anyone know if this is planned for the future?

I haven't even bothered with tone mapping yet. What I will make up for. I spent the weekend setting up the NGINX+LE for Jellyfin :-)

I definitely want an OSS on my box (on the TV) and not a closed box like nVIDIA Shield. Even if the temptation is great. I planned to install a correspondingly powerful nVIDIA card in the server and to keep the minicomputers on the TV's small. But that doesn't make any sense :-(

**Comment by pixel24** [comment]
Hey, thanks for the tip about JellyfinCon. It runs much better that way. Server is well equipped.

Now when I watch a 4K movie I have ~10% CPU load on the server (VM), almost nothing on the Kodi client.

However, I have (minimal) very short stutters in the film.

It can't be because of the network speed. I can read/write from the client to the server at almost 100MB/second.

There are some settings in the JellyfinCon, but I can't find any documentation about them. Can I still optimize something here?
